### PR TITLE
DEV: Update docs config

### PR DESCRIPTION
**Pull request by adehad**
# Summary of changes:
1. [Added] Creates a tox env for docs generating
2. [Subjective, can be removed] Change theme to `furo` due to modern look and better maintenance
3. [Changed] conf.py to prepare for https://github.com/py-pdf/PyPDF2/discussions/1227

# Points of Note
There are about 250 warnings in the autogenerated docs, mainly this is because there are modules prefixed with underscores that are not autogenerated by default. Happy to further add to this PR to address these changes with your guidance, otherwise happy to make a separate PR.

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -15,32 +15,60 @@
 
 sys.path.insert(0, os.path.abspath("."))
 sys.path.insert(0, os.path.abspath("../"))
+import PyPDF2 as py_pkg
```
Comment (MartinThoma) [inline docs/conf.py:18]:
```suggestion
import PyPDF2 as py_pkg  # noqa
```

**Comment by MartinThoma** [comment]
Why did you remove the bat file?

**Comment by adehad** [comment]
> Why did you remove the bat file?

@MartinThoma As the `tox -e docs` is platform agnostic it obsoletes both

**Comment by adehad** [comment]
Managed to shave off around 80 warnings by ignoring the `io` module references. For some reason they are picked up as being under the `_io` module by sphinx and then complains that the module doesn't exist

**Review by MartinThoma — CHANGES_REQUESTED** [review]
When I run `tox -e docs` it doesn't seem to generate anything. Besides several warnings, it generates:

```
copying static files... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... done
build finished with problems, 165 warnings.
ERROR: InvocationError for command /home/moose/Github/py-pdf/PyPDF2/.tox/docs/bin/sphinx-build -a -n -v -W --keep-going -b html --color -d .tox/docs_doctree docs/ .tox/docs_out (exited with code 1)
__________________________________________________________________________________________________________________________________________________________ summary ___________________________________________________________________________________________________________________________________________________________
ERROR:   docs: commands failed
```

At the moment, I think this PR would break the documentation generation.

Did I do something wrong?

**Comment by MartinThoma** [comment]
Regarding the switch to `furo`, I need to compare examples. I make some screenshots on my laptop as well as with the S20 size:

## readthedocs

Start page:

image


Start page (S20):

image

Navbar (S20):

image


Extract Text:

image


Extract Text (S20):

image

PdfFileReader:

image


PdfFileReader (S20):

image


## furo

Start page:

image

Extract Text:

image

PdfFileReader:

image

Start Page (S20):

image

Navbar (S20):

image


PdfFileReader (S20):

image

**Comment by MartinThoma** [comment]
I like `sphinx_rtd_theme` better than `furo`. I also don't see any objective reason to switch (e.g. both look fine on mobile)

**Comment by adehad** [comment]
Thanks for the review @MartinThoma 

> When I run `tox -e docs` it doesn't seem to generate anything. Besides several warnings, it generates:


The output docs should be in the folder
`.tox/docs_out` but happy to change. 


> At the moment, I think this PR would break the documentation generation.
> 
> Did I do something wrong?

The command I used treats warnings as errors, (which is usually what you want when running in CI, so contributors don't introduce more issues). So possible for me to remove that flag due to the current docs state. 




Also happy to move back to the rtd theme, the issue I've had with it is that: it doesn't get updated as regularly. So we had some issues when updating sphinx for example. Looking at the project now perhaps they have gotten some maintenance back in.

**Comment by adehad** [comment]
hmm, just changing to draft as the docs build is running into an error with docutils

**Comment by adehad** [comment]
false alarm it seems! Builds fine for me now, might be because of the cache from the previous docs build

**Comment by MartinThoma** [comment]
Sorry it took that long. I'm ok with the tox changes, but please keep the Makefile / the make.bat:

1. I don't know if readthedocs would work with the current PR. Probably there is a way to make it work, but I don't have the energy to figure that out
2. Developers might want to build the docs locally (hence keeping also the bat file). I would be fine with removing the bat file if you added instructions to `docs/dev/documentation.md` how to generate the documentation with tox.

**Comment by adehad** [comment]
@MartinThoma sounds good. Regarding the readthedocs, I can include a config file that can work. I can base it off work on another project I maintain:
https://github.com/pycontribs/jira/blob/main/.readthedocs.yml

It may also be helpful to configure the readthedocs app, so that this PR will include a link to the docs (which will help confirm it is working)

**Comment by MartinThoma** [comment]
> I can include a config file that can work.

Currently the docs build just fine. Why should we change it?

> It may also be helpful to configure the readthedocs github actions app, so that this PR will include a link to the docs based on these changes (which will help confirm it is working)

What does it do / why is that a good idea?

**Comment by adehad** [comment]
> > I can include a config file that can work.
> 
> Currently the docs build just fine. Why should we change it?

My experience suggests using a "configuration as code" approach is more maintainable. 
One example might be: contributors being able to suggest a fix for broken docs builds without them needing admin rights in readthedocs.

> 
> > It may also be helpful to configure the readthedocs github actions app, so that this PR will include a link to the docs based on these changes (which will help confirm it is working)
> 
> What does it do / why is that a good idea?

**What does it do**: For every pull request it builds the docs, and provides a preview link.
**Why is that a good idea:** I found it a good way to check how a contribution will change the docs and also refer a contributor to how the docs look after their change.

**Comment by MartinThoma** [comment]
> configuration as code

Sounds good :+1: Please go ahead with adding that :-) 

As this PR already takes a while,  I would suggest to make small/single changes. That makes it a lot easier for me to test + decide to merge something. Adding a config file for read-the-docs might be a single simple PR.

**Comment by MartinThoma** [comment]
> For every pull request it builds the docs, and provides a preview link.

That sounds amazing :heart_eyes:

**Comment by adehad** [comment]
Cool, I've tidied up this to just be for tox changes, will leave the config for another PR

**Comment by adehad** [comment]
Doc build failed on the readthedocs, but not reported yet
I will need to drop changes to `requirements/docs.txt` for the build to pass most likely

**Comment by MartinThoma** [comment]
This is open for a long time now and I feel like it will never finish.

The current way of building the docs works, the current PR does not. Hence I close it.

Some of the changes you proposed were added in between. Thank you for the time and effort you put into it :pray:

**Pull request closed without merging**